File: lintcode/python/1377_find_substring.py
#!/usr/bin/python -t

class Solution:
    """
    @param str: The string
    @param k: The length of the substring
    @return: The answer
    """
    def findSubstring(self, str, k):
        # Write your code here
        ch_set = set()
        word_set = set()
        
        n = len(str)
        start, end = 0, 0
        # ch_set.add(str[0])
        
        count = 0
        while end < n:
            if str[end] not in ch_set:
                ch_set.add(str[end])
                end += 1
                
                if len(ch_set) == k:
                    word = str[start:end]
                    if word not in word_set:
                        word_set.add(word)
                        count += 1
                    ch_set.remove(str[start])
                    start += 1
            else:
                ch_set.remove(str[start])
                start += 1
                
        return count

File: lintcode/golang/1377_find_substring.go
/**
 * @param str: The string
 * @param k: The length of the substring
 * @return: The answer
 */
func findSubstring (str string, k int) int {
    // Write your code here
    n := len(str)
    if n == 0  || n < k{
        return 0
    }

    gMap := make(map[string]bool)
    lMap := make(map[string]int)
    l, r := 0, 0
    subStr := ""
    ret := 0
    for r < n {
        subStr = subStr + string(str[r])
        lMap[string(str[r])] += 1
        for len(lMap) > k || lMap[string(str[r])] > 1 {
            lMap[string(str[l])]--
            if lMap[string(str[l])] == 0 {
                delete(lMap, string(str[l]))
            }
            l++
            subStr = subStr[1:]
        }
        if len(lMap) == k && gMap[subStr] == false {
            gMap[subStr] = true
            ret++
        }
        r++
    }

    return ret
}

